btx/misc/pipca_visuals.py

```diff
@@ -0,0 +1,283 @@
+import numpy as np
+import h5py
+
+from btx.interfaces.ipsana import (
+    PsanaInterface,
+    retrieve_pixel_index_map,
+    assemble_image_stack_batch,
+)
+
+import holoviews as hv
+hv.extension('bokeh')
+from holoviews.streams import Params
+
+import panel as pn
+pn.extension(template='fast')
+pn.state.template.param.update()
+import panel.widgets as pnw
+
+def display_dashboard(filename):
+    """
+    Displays an interactive dashboard with a PC plot, scree plot, 
+    and intensity heatmaps of the selected source image as well as 
+    its reconstructed image using the model obtained by pipca.
+    """

```

Comment:
@ThomasBnon will work on this.

Reply:
I added it

Reply:
To summarize:

1. First, I added the filename as a parameter of Class PiPCA: 

```
def __init__(
        self,
        exp,
        run,
        det_type,
        start_offset=0,
        num_images=10,
        num_components=10,
        batch_size=10,
        priming=False,
        downsample=False,
        bin_factor=2,
        output_dir="",
        filename='pipca.model_h5',
    )
```
2. Instead of creating a get_model(self) function to gather U, S, V, mu and var in one function, I created individual gather functions:
```
def gather_U(self):
        """
        Gather and return the U_tot variable.
        """
        if self.rank == 0:
            U_tot = np.empty(self.num_features * self.num_components)

        else:
            U_tot = None

        start_indices = self.split_indices[:-1]

        self.comm.Gatherv(
            self.U.flatten(),
            [
                U_tot,
                self.split_counts * self.num_components,
                start_indices * self.num_components,
                MPI.DOUBLE,
            ],
            root=0,
        )

        if self.rank == 0:
            U_tot = np.reshape(U_tot, (self.num_features, self.num_components))

        return U_tot

    def gather_mu(self):
        """
        Gather and return the mu_tot variable.
        """
        if self.rank == 0:
            mu_tot = np.empty((self.num_features, 1))
        else:
            mu_tot = None

        start_indices = self.split_indices[:-1]
        self.comm.Gatherv(
            self.mu,
            [
                mu_tot,
                self.split_counts * self.num_components,
                start_indices,
                MPI.DOUBLE,
            ],
            root=0,
        )
        return mu_tot

    def gather_var(self):
        """
        Gather and return the var_tot variable.
        """
        if self.rank == 0:
            var_tot = np.empty((self.num_features, 1))
        else:
            var_tot = None

        start_indices = self.split_indices[:-1]
        self.comm.Gatherv(
            self.total_variance,
            [
                var_tot,
                self.split_counts * self.num_components,
                start_indices,
                MPI.DOUBLE,
            ],
            root=0,
        )
        return var_tot
```
3. I renamed variables accordingly to understand mathematics behind it.
4. To unpack variables from pipca_model.h5 file, I modified the function unpack_pipca_model_file(filename):
```
def unpack_pipca_model_file(filename):
    """
    Reads PiPCA model information from h5 file and returns its contents

    Parameters
    ----------
    filename: str
        name of h5 file you want to unpack
        
    Returns
    -------
    data: dict
        A dictionary containing the extracted data from the h5 file.
    """
    data = {}
    with h5py.File(filename, 'r') as f:
        data['exp'] = str(np.asarray(f.get('exp')))[2:-1]
        data['run'] = int(np.asarray(f.get('run')))
        data['det_type'] = str(np.asarray(f.get('det_type')))[2:-1]
        data['start_img'] = int(np.asarray(f.get('start_offset')))
        data['loadings'] = np.asarray(f.get('loadings'))
        data['U'] = np.asarray(f.get('U'))
        data['S'] = np.asarray(f.get('S'))
        data['V'] = np.asarray(f.get('V'))

    return data
```
5. For the function display_eigenimages, the colormap used is now plasma instead of 'BrBG':
```
opts = dict(width=400, height=300, cmap='plasma', colorbar=True,
                    symmetric=True, shared_axes=False, toolbar='above')
```
6. To improve Numpy efficiency, I replaced 
```
 #img_source = None
    #min_diff = None
    #square_diff = None
    
    #for i, (xv, yv) in enumerate(zip(PCx_vector, PCy_vector)):    
        #square_diff = (x - xv) ** 2 + (y - yv) ** 2
        #if (min_diff is None or square_diff < min_diff):
            #min_diff = square_diff
            #img_source = i
```
with
```
    square_diff = (PCx_vector - x)**2 + (PCy_vector - y)**2
    img_source = square_diff.argmin()
```
7. To anticipate failure mechanism in display_error_plots, the code can be modified as described below:
```
try:
   assert num_images * num_features < 4e8, 'image and feature dimensions are too high to test errors'
   assert num_images == num_components, 'number of images must equal the number of components to compare with np.linalg.svd() output'
except AssertionError as e:
   print(f"Assertion Error: {e}")
```
8. To build the heatmap in pipca_visuals.py, I replaced:
```
 bin_y_pixels, bin_x_pixels = binned_img.shape
    rows = np.tile(np.arange(bin_y_pixels).reshape((bin_y_pixels, 1)), bin_x_pixels).flatten()
    cols = np.tile(np.arange(bin_x_pixels), bin_y_pixels)

    hm_data = np.stack((rows, cols, binned_img.flatten()))
    hm_data = hm_data.T.reshape((bin_y_pixels * bin_x_pixels, 3))
```
with:

```
# Create hm_data array for heatmap
bin_x_pixels, bin_y_pixels = binned_img.shape
rows = np.tile(np.arange(bin_x_pixels).reshape((bin_x_pixels, 1)), bin_y_pixels).flatten()
cols = np.tile(np.arange(bin_y_pixels), bin_x_pixels)

hm_data = np.stack((rows, cols, binned_img.flatten())).T
```
9. I renamed variables using common conventions:
```
    y_pixels, x_pixels = img.shape
    bin_factor_y = int(y_pixels / max_pixels)
    bin_factor_x = int(x_pixels / max_pixels)

    while y_pixels % bin_factor_y != 0:
        bin_factor_y += 1
    while x_pixels % bin_factor_x != 0:
        bin_factor_x += 1

    img = img.reshape((y_pixels, x_pixels))
    binned_img = img.reshape(int(y_pixels / bin_factor_y),
                            bin_factor_y,
                            int(x_pixels / bin_factor_x),
                            bin_factor_x).mean(-1).mean(1)

    # Create hm_data array for heatmap
    bin_y_pixels, bin_x_pixels = binned_img.shape
    rows = np.tile(np.arange(bin_y_pixels).reshape((bin_y_pixels, 1)), bin_x_pixels).flatten()
    cols = np.tile(np.arange(bin_x_pixels), bin_y_pixels)

    hm_data = np.stack((rows, cols, binned_img.flatten()))
    hm_data = hm_data.T.reshape((bin_y_pixels * bin_x_pixels, 3))
```
10. I used f-string syntax:
```
opts = dict(width=400, height=300, color='Image', cmap='rainbow', colorbar=True,
                    show_grid=True, shared_axes=False, toolbar='above', tools=['hover'])
        scatter = hv.Points(scatter_data, kdims=[PCx, PCy], vdims=['Image'], 
                            label=f"{PCx.title()} vs {PCy.title()}").opts(**opts)
```
11. Still on duplication, I also modified update_V:
```
def update_V(self, X, S):
        """
        Updates current V with shape (n x q) to shape (n + m, q)
        based on the previous and updated U and S.

        Parameters
        ----------
        X : ndarray, shape (_ x m)
            sliced image batch on self.rank
        S : ndarray, shape (q,)
            singular values of the updated model
        """
        num_incorporated_images = self.num_incorporated_images
        batch_number = self.batch_number

        # Gather all X and U across all ranks
        with TaskTimer(self.task_durations, "V - gather X and U"):
            X_tot = self.gather_X(X)
            U_tot = self.gather_U()

        V = None  # Initialize V

        self.comm.Barrier()

        with TaskTimer(self.task_durations, "V - compute updated V"):
            if num_incorporated_images > 0:
                # U_prev and S_prev aren't instantiated on the first primed batch
                if self.priming and batch_number == 0:
                    self.U_prev = U_tot
                    self.S_prev = S

                # Update each previous V_i from the previous batches
                V = self.V
                B = np.diag(self.S_prev) @ self.U_prev.T @ U_tot @ np.linalg.inv(np.diag(S))
                for i in range(batch_number):
                    start, end = self.batch_indices[i], self.batch_indices[i + 1]
                    V[start:end] = V[start:end] @ B

                # Compute new V_m from the current batch with standard SVD
                V_m = X_tot.T @ U_tot @ np.linalg.inv(np.diag(S))
                V = np.concatenate((V, V_m))
            else:
                # Instantiate self.V with standard SVD
                V = X_tot.T @ U_tot @ np.linalg.inv(np.diag(S))

        self.comm.Barrier()

        with TaskTimer(self.task_durations, "V - bcast V"):
            self.comm.Bcast(V, root=0)

        self.V = V
        self.U_prev = U_tot
        self.S_prev = S
```
and created:
```
    def gather_X(self, X):
        
        Gather and return the X_tot variable.
        
        _, m = X.shape
        if self.rank == 0:
            X_tot = np.empty((self.num_features, m))
        else:
            X_tot = None

        start_indices = self.split_indices[:-1]

        self.comm.Gatherv(
            X.flatten(),
            [
                X_tot,
                self.split_counts * m,
                start_indices * m,
                MPI.DOUBLE,
            ],
            root=0,
        )

        if self.rank == 0:
            X_tot = np.reshape(X_tot, (self.num_features, m))

        return X_tot
```
12. I also described functions correctly (Parameters, outputs etc.). 
To compare with previous PR from Russel, here is the dashboard I obtained after all these modifications with exactly the same analysis parameters:
![Capture](https://github.com/russell-marasigan/btx/assets/148398571/e4667e15-5697-460b-9eb3-48626398b1ae)

Compared to Russel dashboard:
![263874791-65acb09a-3f2c-4afd-b58e-cac461adbf82](https://github.com/russell-marasigan/btx/assets/148398571/2b6373e1-d301-4e58-9b83-451e574225e4)